### PR TITLE
Add regionMapping applyReplacements from v7

### PR DESCRIPTION
**Pull request by nf-s**
### Add regionMapping applyReplacements from v7

This enables region mapping `dataReplacements` to work.

Also I started a typescript file to move region mapping stuff across - https://github.com/TerriaJS/terriajs/issues/4715

### Checklist

-   [x] Tests will wait until more Typescript has been written
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

File: lib/Map/RegionProviderTs.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ */
+export function applyReplacements(
+  regionProvider: RegionProvider,
+  s: string | number,
+  replacementsProp: "dataReplacements" = "dataReplacements"
+): string | undefined {
+  if (!isDefined(s)) {
+    return undefined;
+  }
+  var r: string;
```
Comment (KeyboardSounds) [inline lib/Map/RegionProviderTs.ts]: `var` -> `let`

File: lib/Map/RegionProviderTs.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  if (typeof s === "number") {
+    r = String(s);
+  } else {
+    r = s.toLowerCase().trim();
+  }
+  var replacements = regionProvider[replacementsProp];
```
Comment (KeyboardSounds) [inline lib/Map/RegionProviderTs.ts]: `var` -> `let`

File: lib/Map/RegionProviderTs.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * Apply an array of regular expression replacements to a string. Also caches the applied replacements in regionProvider._appliedReplacements.
+ * @private
+ * @param {RegionProvider} regionProvider The RegionProvider instance.
+ * @param {String} s The string.
+ * @param {String} replacementsProp Name of a property containing [ [ regex, replacement], ... ], where replacement is a string which can contain '$1' etc.
+ */
+export function applyReplacements(
+  regionProvider: RegionProvider,
+  s: string | number,
+  replacementsProp: "dataReplacements" = "dataReplacements"
```
Comment (KeyboardSounds) [inline lib/Map/RegionProviderTs.ts]: The type here doesn't make sense. If the only possible value of replacementsProp is `"dataReplacements"`, then why make it a parameter at all? I'm guessing it should be a string.

**Review by KeyboardSounds — CHANGES_REQUESTED** [review]

File: lib/Table/TableColumn.ts [hunk 4]
```diff
@@ -313,7 +314,10 @@ export default class TableColumn {
   ): string | null {
     // TODO: validate that the rowValue is actually a valid region, if possible.
     // TODO: implement replacements
```
[the remark this answers is not in the document]
Comment (nf-s) [reply lib/Table/TableColumn.ts]: there are a few other replacements which haven't been added, so i'll modify the comment

File: lib/Map/RegionProviderTs.ts (hunk 3 above)
Comment (nf-s) [reply lib/Map/RegionProviderTs.ts]: Type should have been ` "dataReplacements"   | "serverReplacements"  | "disambigDataReplacements"`

**Review by KeyboardSounds — APPROVED** [review]

**Pull request merged**